src/main/scala/chisel3/tester/BackendInterface.scala

```diff
@@ -131,15 +131,64 @@ trait BackendInstance[T <: MultiIOModule] extends BackendInterface {
 /** Interface into the testing environment, like ScalaTest
   */
 trait TestEnvInterface {
+
+  protected val batchedFailures: mutable.ArrayBuffer[Exception] = new mutable.ArrayBuffer
+
+  def topFileName: Option[String]
+
   /** Logs a tester failure at this point.
     * Failures queued until the next checkpoint.
     */
-  def testerFail(msg: String): Unit
+  def testerFail(msg: String): Unit = {
+    batchedFailures += new Exception(s"$msg")
+  }
+
+  protected def getExpectDetailedTrace(trace: Seq[StackTraceElement], inFile: String): String = {
+    val fullTrace = Context().backend.getParentTraceElements ++ trace
+
+    // In the threading case, this needs to be overridden to trace through parent threads
+    val lineNumbers = fullTrace.collect {
+      case ste if ste.getFileName == inFile => ste.getLineNumber
+    }.mkString(", ")
+    if (lineNumbers.isEmpty) {
+      s" (no lines in $inFile)"
+    } else {
+      s" (lines in $inFile: $lineNumbers)"
+    }
+  }
+
   /** Expect a specific value on a wire, calling testerFail if the expectation isn't met.
     * Failures queued until the next checkpoint.
     */
-  def testerExpect(expected: Any, actual: Any, signal: String, msg: Option[String]): Unit
+  def testerExpect(expected: Any, actual: Any, signal: String, msg: Option[String]): Unit = {
+    if (expected != actual) {
+      val appendMsg = msg match {
+        case Some(_) => s": $msg"
+        case _ => ""
+      }
+
+      val trace = new Throwable
+      val expectStackDepth = trace.getStackTrace.indexWhere(ste =>
+        ste.getClassName == "chisel3.tester.package$testableData" && ste.getMethodName == "expect")
+      require(expectStackDepth != -1,
+        s"Failed to find expect in stack trace:\r\n${trace.getStackTrace.mkString("\r\n")}")
+
+      val trimmedTrace = trace.getStackTrace.drop(expectStackDepth + 2)
+      val detailedTrace = topFileName.map(getExpectDetailedTrace(trimmedTrace.toSeq, _)).getOrElse("")
+
+      batchedFailures += new TestFailedException(
```

Comment:
You should throw a testers2-specific exception (eg ExpectMismatchException) instead of reusing Scalatest infrastructure for a trait that's not supposed to be Scalatest specific

Reply:
Ok, making it into FailedExpectException